File: hivemind/cluster/router.py
"""Task routing with affinity: memory, tool cache, load."""

from hivemind.cluster.node_info import NodeInfo
from hivemind.types.task import Task


def _parse_version(version: str) -> tuple[int, int]:
    """Return (major, minor) from version string like 1.10 or 1.9.0."""
    parts = version.replace("-", ".").split(".")
    try:
        major = int(parts[0]) if parts else 0
        minor = int(parts[1]) if len(parts) > 1 else 0
    except (ValueError, IndexError):
        major, minor = 0, 0
    return (major, minor)


def _clamp(value: float, lo: float, hi: float) -> float:
    return max(lo, min(hi, value))


class TaskRouter:
    """Route ready tasks to the best available worker. Scoring: memory_affinity, tool_affinity, load_score."""

    # Weights
    MEMORY_WEIGHT = 0.35
    TOOL_WEIGHT = 0.25
    LOAD_WEIGHT = 0.40

    def __init__(self, controller_version: str = "1.9.0") -> None:
        self._controller_version = controller_version
        self._ctrl_major, self._ctrl_minor = _parse_version(controller_version)
        self._round_robin_index: int = 0  # spread tasks when scores tie

    def route(
        self,
        task: Task,
        workers: list[NodeInfo],
        worker_stats: dict[str, dict],
    ) -> NodeInfo | None:
        """Score all workers; return highest-scored; when scores tie, round-robin so tasks spread."""
        if not workers:
            return None
        eligible = [w for w in workers if self._version_compatible(w)]
        if not eligible:
            return None
        scored = [(self._score(task, w, worker_stats.get(w.node_id, {})), w) for w in eligible]
        best_score = max(s[0] for s in scored)
        if best_score < 0:
            return None
        # Treat scores within 1e-6 as tied so we round-robin and spread (avoids one worker getting all + 429)
        eps = 1e-6
        tied = [w for s, w in scored if abs(s - best_score) <= eps]
        # Round-robin among tied workers so multiple tasks go to different workers
        idx = self._round_robin_index % len(tied)
        self._round_robin_index += 1
        return tied[idx]

    def _version_compatible(self, worker: NodeInfo) -> bool:
        maj, min_ = _parse_version(worker.version)
        return (maj == self._ctrl_major and min_ == self._ctrl_minor)

    def _score(
        self,
        task: Task,
        worker: NodeInfo,
        stats: dict,
    ) -> float:
        memory_affinity = self._memory_affinity(task, stats)
        tool_affinity = self._tool_affinity(task, stats)
        load_score = self._load_score(worker, stats)
        return (
            memory_affinity * self.MEMORY_WEIGHT
            + tool_affinity * self.TOOL_WEIGHT
            + load_score * self.LOAD_WEIGHT
        )

    def _memory_affinity(self, task: Task, stats: dict) -> float:
        deps = list(task.dependencies or [])
        if not deps:
            return 0.0
        completed_here = list(stats.get("completed_task_ids", []))
        hit = sum(1 for d in deps if d in completed_here)
        return hit / len(deps)

    def _tool_affinity(self, task: Task, stats: dict) -> float:
        tools_needed = self._tools_for_task(task)
        if not tools_needed:
            return 0.0
        cached = set(stats.get("cached_tools", []))
        hit = sum(1 for t in tools_needed if t in cached)
        return hit / len(tools_needed)

    def _tools_for_task(self, task: Task) -> list[str]:
        try:
            from hivemind.tools.selector import get_tools_for_task
            from hivemind.tools.scoring import get_default_score_store
            score_store = get_default_score_store()
        except Exception:
            score_store = None
        try:
            tools = get_tools_for_task(
                task.description or "",
                role=getattr(task, "role", None),
                score_store=score_store,
            )
            return [t.name for t in tools]
        except Exception:
            return []

    def _load_score(self, worker: NodeInfo, stats: dict) -> float:
        active = int(stats.get("active_tasks", 0))
        avg_duration = float(stats.get("avg_task_duration_seconds", 0.0))
        max_workers = worker.max_workers or 1
        weighted_load = active * avg_duration
        load_ratio = weighted_load / (max_workers * 60.0)
        return 1.0 - _clamp(load_ratio, 0.0, 1.0)

File: worker/src/router.rs
//! Task routing — memory affinity, tool affinity, load score.

use std::collections::HashMap;

use crate::types::{NodeInfo, Task};

fn parse_version(version: &str) -> (u32, u32) {
    let binding = version.replace('-', ".");
    let parts: Vec<&str> = binding.split('.').collect();
    let major = parts.first().and_then(|s| s.parse().ok()).unwrap_or(0);
    let minor = parts.get(1).and_then(|s| s.parse().ok()).unwrap_or(0);
    (major, minor)
}

fn clamp(value: f64, lo: f64, hi: f64) -> f64 {
    value.max(lo).min(hi)
}

/// Route ready task to best worker. Scoring: memory_affinity, tool_affinity, load_score.
pub struct TaskRouter {
    controller_version: String,
    round_robin_index: usize,
}

impl TaskRouter {
    pub const MEMORY_WEIGHT: f64 = 0.35;
    pub const TOOL_WEIGHT: f64 = 0.25;
    pub const LOAD_WEIGHT: f64 = 0.40;

    pub fn new(controller_version: impl Into<String>) -> Self {
        Self {
            controller_version: controller_version.into(),
            round_robin_index: 0,
        }
    }

    pub fn route<'a>(
        &mut self,
        task: &Task,
        workers: &'a [NodeInfo],
        worker_stats: &HashMap<String, serde_json::Value>,
    ) -> Option<&'a NodeInfo> {
        if workers.is_empty() {
            return None;
        }
        let (ctrl_major, ctrl_minor) = parse_version(&self.controller_version);
        let eligible: Vec<&'a NodeInfo> = workers
            .iter()
            .filter(|w| {
                let (maj, min) = parse_version(&w.version);
                maj == ctrl_major && min == ctrl_minor
            })
            .collect();
        if eligible.is_empty() {
            return None;
        }
        let empty = serde_json::Map::new();
        let scored: Vec<(f64, &'a NodeInfo)> = eligible
            .iter()
            .map(|w| {
                let stats = worker_stats
                    .get(&w.node_id)
                    .and_then(|v| v.as_object())
                    .unwrap_or(&empty);
                let score = self.score(task, w, stats);
                (score, *w)
            })
            .collect();
        let best = scored.iter().map(|(s, _)| *s).fold(0.0_f64, f64::max);
        if best < 0.0 {
            return None;
        }
        let eps = 1e-6;
        let tied: Vec<&'a NodeInfo> = scored
            .iter()
            .filter(|(s, _)| (s - best).abs() <= eps)
            .map(|(_, w)| *w)
            .collect();
        let idx = self.round_robin_index % tied.len();
        self.round_robin_index += 1;
        Some(tied[idx])
    }

    fn score(
        &self,
        task: &Task,
        worker: &NodeInfo,
        stats: &serde_json::Map<String, serde_json::Value>,
    ) -> f64 {
        let memory = memory_affinity(task, stats);
        let tool = tool_affinity(task, stats);
        let load = load_score(worker, stats);
        memory * Self::MEMORY_WEIGHT + tool * Self::TOOL_WEIGHT + load * Self::LOAD_WEIGHT
    }
}

fn memory_affinity(task: &Task, stats: &serde_json::Map<String, serde_json::Value>) -> f64 {
    let deps = &task.dependencies;
    if deps.is_empty() {
        return 0.0;
    }
    let completed: Vec<String> = stats
        .get("completed_task_ids")
        .and_then(|v| v.as_array())
        .map(|a| {
            a.iter()
                .filter_map(|v| v.as_str().map(String::from))
                .collect()
        })
        .unwrap_or_default();
    let hit = deps.iter().filter(|d| completed.contains(d)).count();
    hit as f64 / deps.len() as f64
}

fn tool_affinity(_task: &Task, stats: &serde_json::Map<String, serde_json::Value>) -> f64 {
    let cached: Vec<String> = stats
        .get("cached_tools")
        .and_then(|v| v.as_array())
        .map(|a| {
            a.iter()
                .filter_map(|v| v.as_str().map(String::from))
                .collect()
        })
        .unwrap_or_default();
    if cached.is_empty() {
        return 0.0;
    }
    // Rust worker doesn't have Python tool selector; use 0 or 1.0 to not penalize
    0.0
}

fn load_score(worker: &NodeInfo, stats: &serde_json::Map<String, serde_json::Value>) -> f64 {
    let active = stats
        .get("active_tasks")
        .and_then(|v| v.as_u64())
        .unwrap_or(0) as f64;
    let avg_duration = stats
        .get("avg_task_duration_seconds")
        .and_then(|v| v.as_f64())
        .unwrap_or(0.0);
    let max_workers = worker.max_workers as f64;
    let weighted_load = active * avg_duration;
    let load_ratio = weighted_load / (max_workers * 60.0).max(1.0);
    1.0 - clamp(load_ratio, 0.0, 1.0)
}
